plugin-server/src/cdp/services/job-queue/job-queue-delay.ts

```diff
@@ -0,0 +1,204 @@
+        await this.kafkaConsumer.connect(async (messages) => {
+            const { backgroundTask } = await this.consumeKafkaBatch(messages)
+            return { backgroundTask }
+        })
+    }
+
+    public async stopConsumer() {
+        await this.kafkaConsumer?.disconnect()
+    }
+
+    public async stopProducer() {
+        await this.kafkaProducer?.disconnect()
+    }
+
+    public isHealthy() {
+        return this.kafkaConsumer?.isHealthy() ?? false
+    }
+
+    private getKafkaProducer(): KafkaProducerWrapper {
+        if (!this.kafkaProducer) {
+            throw new Error('KafkaProducer not initialized')
+        }
+        return this.kafkaProducer
+    }
+
+    private getHeaderValue(headers: any[] | undefined, key: string): string | undefined {
+        if (!headers || !Array.isArray(headers)) {
+            return undefined
+        }
+
+        for (const header of headers) {
+            if (header[key]) {
+                // Convert Uint8Array to string
+                return Buffer.from(header[key]).toString('utf8')
+            }
+        }
+        return undefined
+    }
+
+    private async consumeKafkaBatch(messages: Message[]): Promise<{ backgroundTask: Promise<any> }> {
+        if (messages.length === 0) {
+            return await this.consumeBatch([])
+        }
+
+        logger.info('🔁', `${this.name} - Consuming batch`, { messageCount: messages.length })
+
+        const maxDelayMs = getDelayByQueue(this.queue)
+
+        for (const message of messages) {
+            try {
+                const returnTopic = this.getHeaderValue(message.headers, 'returnTopic') as CyclotronJobQueueKind
+                const queueScheduledAt = this.getHeaderValue(message.headers, 'queueScheduledAt')
+
+                if (!returnTopic || !queueScheduledAt) {
+                    logger.warn('Missing required headers', {
+                        returnTopic,
+                        queueScheduledAt,
+                        messageKey: message.key,
+                    })
+                    this.kafkaConsumer?.offsetsStore([
+                        {
+                            ...message,
+                            offset: message.offset + 1,
+                        },
+                    ])
+                    continue
+                }
+
+                const now = new Date().getTime()
+                const scheduledTime = new Date(queueScheduledAt)
+                let delayMs = Math.max(0, scheduledTime.getTime() - now)
+                const waitTime = Math.min(delayMs, maxDelayMs)
+
+                logger.info(
+                    '🔁',
+                    `${this.name} - Waiting for ${waitTime}ms before processing ${messages.indexOf(message) + 1}/${messages.length} invocation ${message.key}`
+                )
+
+                delayMs -= waitTime
+
+                await new Promise((resolve) => setTimeout(resolve, waitTime))
```

Comment:
We need to be able to cleanly cancel this, for example in the case of a shutdown event. Not sure the best way of going about this but given that we are in charge of the offset store its _probably_ safe to just have some sort of special handler that checks every second for whether we are rebalancing or shutting down and then just throws or something